File: cpp_solutions/5_longestPalindromicSubstring.cpp
#include<string>
#include<stdint.h>
using namespace std;

/*
Given a string S, find the longest palindromic substring in S. 
You may assume that the maximum length of S is 1000, 
and there exists one unique longest palindromic substring.
wiki:
http://www.wikiwand.com/en/Longest_palindromic_substring
*/
class Solution {
public:
	string longestPalindrome(string s) {
		int len = s.length();
		if (len <= 1)
		{
			return s;
		}
		else
		{
			int maxLen = 1;
			int begin = 0;
			//string longestString = s[0];
			for (int i = 0; i < len-1; i++)
			{
				getLongest(s, i, i, begin, maxLen);
				getLongest(s, i, i + 1, begin, maxLen);
			}
			return s.substr(begin, maxLen);
		}
	}

	void getLongest(string s, int left, int right, int& begin, int& maxLen)
	{
		int len = 0;
		while (left>=0 && right <= s.length() && s[left]==s[right])
		{
			if (left == right)
			{
				len = len + 1;
			}
			else
			{
				len = len + 2;
			}
			left--;
			right++;
		}
		if (len > maxLen)
		{
			maxLen = len;
			begin = left + 1;
		}
	}

};

int main()
{
	int c = 11 & 0x1;
	int d = 11 >> 1;
	Solution *s = new Solution();
	string a = "aaaaaaa";
	string b = s->longestPalindrome(a);
	return 0;
}

File: python/5_longestPalindromicSubstring.py
class Solution:
    # @param {string} s
    # @return {string}
    def longestPalindrome(self, s):

        n = len(s)
        if n <= 1:
            return s
        maxLen = 1
        begin = 0

        for i in xrange(n-1):
            left = i - 1
            right = i + 1
            while left >= 0 and s[left]==s[i]:
                left -= 1
            while right < n and s[right]==s[i]:
                right += 1
            l = right - left - 1

            while left>=0 and right<len(s) and s[left] == s[right]:
                l += 2
                left -= 1
                right += 1

            if l > maxLen:
                begin = left + 1
                maxLen = l

        return s[begin:begin+maxLen]

    def longestPalindrome1(self, s):

        n = len(s)
        if n <= 1:
            return s
        maxLen = 1
        begin = 0

        for i in xrange(n-1):
            b, l = self.get(s, i, i)
            if l > maxLen:
                begin = b
                maxLen = l
            b, l = self.get(s, i, i+1)
            if l > maxLen:
                begin = b
                maxLen = l
        return s[begin:begin+maxLen]

    def get(self, s, left, right):
        l = 0
        while left>=0 and right<len(s) and s[left] == s[right]:
            if left == right:
                l += 1
            else:
                l += 2
            left -= 1
            right += 1

        return left+1, l
